File: src/models/AssignmentSubmission.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const assignmentSubmissionSchema = new Schema(
  {
    assignmentId: { type: Schema.Types.ObjectId, ref: 'Assignment', required: true },
    subjectId: { type: Schema.Types.ObjectId, ref: 'Subject', required: true },
    departmentId: { type: Schema.Types.ObjectId, ref: 'Department', required: true },
    studentId: { type: String, ref: 'User', required: true },

    textAnswer: { type: String, default: null },
    mediaId: { type: Schema.Types.ObjectId, ref: 'Media', default: null },
    fileUrl: { type: String, default: null },
    fileName: { type: String, default: null },
    filePublicId: { type: String, default: null },
    fileResourceType: { type: String, default: null },
    fileSize: { type: Number, default: null },

    submittedAt: { type: Date, default: Date.now },

    // null marksObtained == not yet graded (the "isGraded" flag, same as the pre-migration app).
    marksObtained: { type: Number, default: null },
    feedback: { type: String, default: null },
    gradedBy: { type: String, ref: 'User', default: null },
    gradedAt: { type: Date, default: null },
  },
  { timestamps: true }
);

// One submission per student per assignment - resubmitting overwrites (upsert), matching
// the old deterministic "assignmentId_studentUid" doc id.
assignmentSubmissionSchema.index({ assignmentId: 1, studentId: 1 }, { unique: true });

module.exports = mongoose.model('AssignmentSubmission', assignmentSubmissionSchema);

File: app/src/main/java/com/example/uos_lms/core/domain/model/AssignmentSubmission.java
package com.example.uos_lms.core.domain.model;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@Builder(toBuilder = true)
@EqualsAndHashCode
@ToString
public class AssignmentSubmission {
    private final String id;
    private final String assignmentId;
    private final String subjectId;
    private final String studentUid;
    private final String studentName;
    private final String textAnswer;
    private final String fileUrl;
    private final String fileName;
    private final String filePublicId;
    private final String fileResourceType;
    private final Long fileSize;
    @Builder.Default
    private final long submittedAt = 0L;
    private final Integer marksObtained;
    private final String feedback;
    private final String gradedBy;
    private final Long gradedAt;

    public boolean isGraded() {
        return marksObtained != null;
    }
}
